### PR TITLE
[3.1] Fix for block producer stuck in coma state

### PR DESCRIPTION
**Pull request by heifner**
Backports: https://github.com/eosnetworkfoundation/mandel/pull/495 Send handshake for heartbeat/2 if no block (first commit)
* Send handshake message if we have not seen or sent a block for heartbeat/2.
* Always send handshake message even if it has not changed since it was last sent.

Additional changes in second commit:
* Main change is to always start syncing from LIB instead of HEAD.
  * See https://github.com/eosnetworkfoundation/mandel/issues/623#issuecomment-1179463090 
* Change default keepalive interval to 10s. Previous default of 32s seems rather excessive given that block time is 500ms.
* Honor keepalive interval regardless of protocol version since there is nothing that requires a particular protocol version.

### CODE REVIEW

File: plugins/net_plugin/net_plugin.cpp [hunk 1]
```diff
@@ -1150,7 +1150,7 @@ namespace eosio {
 
    // called from connection strand
    void connection::check_heartbeat( tstamp current_time ) {
-      if( protocol_version >= heartbeat_interval && latest_msg_time > 0 ) {
+      if( latest_msg_time > 0 ) {
```
Comment (linh2931) [inline plugins/net_plugin/net_plugin.cpp:1153]: Not related to this PR. The version names of net_plugin should have a prefix of protocol or version. When I first read `protocol_version >= heartbeat_interval`, I thought `protocol_version` and `heartbeat_interval` were different units, how could they compare? Then found out `heartbeat_interval` talked about a version, not an interval value.

File: plugins/net_plugin/net_plugin.cpp [hunk 2]
```diff
@@ -1148,21 +1149,31 @@ namespace eosio {
...
             no_retry = benign_other;
             if( !peer_address().empty() ) {
-               peer_wlog(this, "heartbeat timed out for peer address");
-               close(true);  // reconnect
+               fc_wlog(logger, "heartbeat timed out for peer address ${adr}", ("adr", peer_address()));
+               close(true);
             } else {
-               peer_wlog( this, "heartbeat timed out" );
+               {
+                  std::lock_guard<std::mutex> g_conn( conn_mtx );
```
Comment (linh2931) [inline plugins/net_plugin/net_plugin.cpp:1161]: Just wondering why mutex is required here.

**Review by linh2931 — APPROVED** [review]

File: plugins/net_plugin/net_plugin.cpp (hunk 2 above)
Comment (heifner) [reply plugins/net_plugin/net_plugin.cpp:1161]: `last_handshake_recv` is accessed across threads. Should probably log outside the mutex.

Comment (heifner) [reply plugins/net_plugin/net_plugin.cpp:1161]: Actually this is in a connection strand call, so it should use peer_wlog.

File: plugins/net_plugin/net_plugin.cpp [hunk 3]
```diff
@@ -225,7 +225,7 @@ namespace eosio {
       boost::asio::steady_timer::duration   connector_period{0};
       boost::asio::steady_timer::duration   txn_exp_period{0};
       boost::asio::steady_timer::duration   resp_expected_period{0};
-      std::chrono::milliseconds             keepalive_interval{std::chrono::milliseconds{32 * 1000}};
+      std::chrono::milliseconds             keepalive_interval{std::chrono::milliseconds{10 * 1000}};
```
[the remark this answers is not in the document]
Comment (heifner) [reply plugins/net_plugin/net_plugin.cpp:228]: Done.

File: plugins/net_plugin/net_plugin.cpp (hunk 1 above)
Comment (heifner) [reply plugins/net_plugin/net_plugin.cpp:1153]: Some already had `proto_` so added that to all of them.

**Review by linh2931 — APPROVED** [review]

**Pull request merged**